api/v1alpha4/clusterclass_webhook.go

```diff
@@ -0,0 +1,197 @@
+var _ webhook.Defaulter = &ClusterClass{}
+
+// Default satisfies the defaulting webhook interface.
+func (in *ClusterClass) Default() {
+	// Default all namespaces in the references to the object namespace.
+	if len(in.Spec.Infrastructure.Ref.Namespace) == 0 {
+		in.Spec.Infrastructure.Ref.Namespace = in.Namespace
+	}
+	if len(in.Spec.ControlPlane.Ref.Namespace) == 0 {
+		in.Spec.ControlPlane.Ref.Namespace = in.Namespace
+	}
+	for i := range in.Spec.Workers.MachineDeployments {
+		if len(in.Spec.Workers.MachineDeployments[i].Template.Bootstrap.Ref.Namespace) == 0 {
+			in.Spec.Workers.MachineDeployments[i].Template.Bootstrap.Ref.Namespace = in.Namespace
+		}
+		if len(in.Spec.Workers.MachineDeployments[i].Template.Infrastructure.Ref.Namespace) == 0 {
+			in.Spec.Workers.MachineDeployments[i].Template.Infrastructure.Ref.Namespace = in.Namespace
+		}
+	}
+}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type.
+func (in *ClusterClass) ValidateCreate() error {
+	return in.validate(nil)
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type.
+func (in *ClusterClass) ValidateUpdate(old runtime.Object) error {
+	oldClusterClass, ok := old.(*ClusterClass)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected a ClusterClass but got a %T", old))
+	}
+	return in.validate(oldClusterClass)
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type.
+func (in *ClusterClass) ValidateDelete() error {
+	return nil
+}
+
+func (in *ClusterClass) validate(old *ClusterClass) error {
+	var allErrs field.ErrorList
+
+	// NOTE: ClusterClass and managed topologies are behind ClusterManagedTopologies feature gate flag; the web hook
+	// must prevent creating new objects in case the feature flag is disabled.
+	if !feature.Gates.Enabled(feature.ClusterManagedTopologies) {
+		allErrs = append(
+			allErrs,
+			field.Forbidden(
+				field.NewPath("spec"),
+				"can be set only if ClusterManagedTopologies feature flag is enabled",
+			),
+		)
+	}
+
+	// ensure all the references are within the same namespace
+	if in.Spec.Infrastructure.Ref.Namespace != in.Namespace {
+		allErrs = append(
+			allErrs,
+			field.Invalid(
+				field.NewPath("spec", "infrastructure", "ref", "namespace"),
+				in.Spec.Infrastructure.Ref.Namespace,
+				"must match metadata.namespace",
+			),
+		)
+	}
+	if in.Spec.ControlPlane.Ref.Namespace != in.Namespace {
+		allErrs = append(
+			allErrs,
+			field.Invalid(
+				field.NewPath("spec", "controlPlane", "ref", "namespace"),
+				in.Spec.ControlPlane.Ref.Namespace,
+				"must match metadata.namespace",
+			),
+		)
+	}
+	for _, class := range in.Spec.Workers.MachineDeployments {
+		if class.Template.Bootstrap.Ref.Namespace != in.Namespace {
+			allErrs = append(allErrs,
+				field.Invalid(
+					field.NewPath("spec", "worker", "machineDeployments", "template", "bootstrap", "ref", "namespace"),
```

Comment:
Sorry last one "worker" => "workers" 4x here and 1x in `cluster_webhook.go`

Reply:
Done!